Akka Remote compiling on Scala 3.0

johanandren opened the pull request:
Should allow us to compile and run stream tests on Scala 3 as the next step.

Refs #30243

johanandren committed:
Redundant final

johanandren committed:
Logger class parameter

johanandren committed:
Explicit apply method references

johanandren committed:
Option overload for address not found by compiler?

johanandren committed:
Cyclic import removed

johanandren committed:
Problematic copy overload (deprecated since 2.4) removed.

johanandren committed:
Private constructor when referring to private parameter type

johanandren committed:
Explicit system to Tcp extension

johanandren committed:
Shameless type casting

johanandren committed:
Missing import

johanandren committed:
Queue[ByteString].enqueue binds to overload taking an iterable for some reason

johanandren committed:
Some hoops to avoid typechecks not actually possible (anymore?)

johanandren committed:
Pattern match that can only be null

johanandren committed:
Type parameter not know in pattern match

johanandren committed:
Remove SerialVersionUid for traits

johanandren commented:
Hmm, not entirely sure what happened with the commit history here, I'll anyway have to rebase or something when the upstream PR has been merged.

johanandren committed:
Formatting

johanandren committed:
Mima filters for the removed QuarantinedEvent.copy

johanandren committed:
Unused fields removed

johanandren committed:
Add remoting to travis scala 3 compile

raboof commented on akka-remote/src/main/scala/akka/remote/transport/ThrottlerTransportAdapter.scala:
```diff
@@ -474,7 +474,7 @@ private[transport] class ThrottledAssociation(
   when(WaitOrigin) {
     case Event(InboundPayload(p), ExposedHandle(exposedHandle)) =>
-      throttledMessages = throttledMessages.enqueue(p)
+      throttledMessages = throttledMessages.appended(p: ByteString)
```
apparently `value appended is not a member of scala.collection.immutable.Queue[akka.util.ByteString]` on Scala 2.12, yet :/

johanandren replied:
Dang, and on Scala 3 enqueue binds to the wrong method here (the deprecated one taking Iterable or something like that). I guess that could potentially be a compiler bug though, choosing the wrong overload.

patriknw commented on akka-remote/src/main/scala/akka/remote/serialization/MiscMessageSerializer.scala:
```diff
@@ -485,16 +485,16 @@ class MiscMessageSerializer(val system: ExtendedActorSystem) extends SerializerW
-      if (a.hasHostname) Some(a.getHostname) else None,
-      if (a.hasPort) Some(a.getPort) else None)
+      a.getHostname,
+      a.getPort)
```
is this the same thing? isn't `a.getPort` 0 when not defined and then we will use `Some(0)` instead of `None`?

johanandren replied:
Aha, crap, you are right. 

For the record the problem here was that Scala 3 refuses to bind to the apply overload taking two options last (must be a compiler bug of some sort but didn't dig deeper into that yet).

raboof replied:
that constructor was private, so it looks like dotty was right about that one actually

patriknw reviewed (commented):
reviewed akka-remote/src/main/scala/akka/remote/*

raboof commented:
refs #30243

raboof commented on project/Dependencies.scala:
```diff
@@ -293,14 +293,15 @@ object Dependencies {
-        Test.scalatest)
+        Test.scalatest) ++
+    (if (getScalaVersion() == scala3Version) Seq.empty
+    else Seq(jacksonScala))
```
Sneaky :) - how optional is this dependency in practice?

johanandren replied:
It's specific jackson stuff for Scala SDK types, so I guess a user app could still use Jackson with Scala as long as you don't have options, collections etc in the types you want to serialize. (you likely do). We don't require it for any internals though.

I don't see this as the end state, just a way to get it compiling until there is a jackson scala module for Scala 3

raboof committed:
Use :+ rather than '.appended()'

Because '.appended()' is not available on 2.12 yet

raboof committed:
Open up Address constructor with options

It was used during serialization but apparently this was never a problem
until Scala 3
